docs: Reorganize code-style documentation.

karlstolley opened the pull request:
<!-- Describe your pull request here.-->

**This PR is not yet ready for review.**

This quite large PR restructures code style and conventions documentation with an eye towards reader-friendliness, and is meant to reflect current best practices adhered to by Zulip contributors.

(More description to come.)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -31,7 +31,7 @@ style, fix it (in a separate commit).
 [#development help](https://chat.zulip.org/#narrow/stream/49-development-help).
 
-## Lint tools
+### Make use of the lint tools
```
"Use the linters" might be a better action verb.

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -59,45 +59,50 @@ The Vagrant setup process runs this for you.
 - custom checks (e.g. trailing whitespace and spaces-not-tabs)
 
-### Tests
+### Check your code against automated tests
```
"Use tests to verify logic you write. This section might want to link to the testing philosophy guide and/or

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -71,13 +64,20 @@ The Vagrant setup process runs this for you.
+- custom checks (e.g., trailing whitespace and spaces-not-tabs)
+
+If you'd like, you can set up a local Git commit hook that will lint
+only your changed files each time you commit:
```
I think this section should somehow advertise the ability to run these against individual target files -- it's documented in detail on the page on linters, but a few words on that being a good idea would be valuable here.

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -124,26 +124,23 @@ rules about integrating third-party projects.
+  [integrate Black](https://black.readthedocs.io/en/stable/integrations/editors.html)
   and
   [isort](https://pycqa.github.io/isort/#installing-isorts-for-your-preferred-text-editor)
   with your editor.
```
@andersk should this `isort` advice be replaced with a `ruff` equivalent?

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -152,41 +152,40 @@ all code via our [linter tool](../testing/linters.md) with
-Combine adjacent on-ready functions, if they are logically related.
-
-The best way to build complicated DOM elements is a Handlebars template
+- Combine adjacent on-ready functions, if they are logically related.
```
This advice is probably stale -- we no longer use on-ready functions as a habit at all, in favor of the `ui_init.js` module calling explicit initialize methods for each module.

timabbott commented on docs/contributing/code-style.md:
````diff
@@ -152,41 +152,40 @@ all code via our [linter tool](../testing/linters.md) with
 
 ```js
-<p onclick="select_zerver({{id}})">
+$("#main_div").on("click", ".message_reaction", function (e) {
````
This example is a bit weird, in that these days, we mainly do `$("body")` as the container for this sort of thing.

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -368,9 +369,9 @@ Additional notes:
+to make a bug that's impossible to track down.
 
 You should instead use the `id` function from the `rows` module, as in
 `rows.id(x)`. This returns a number. Even in cases where you do want a
```
I think I've not seen someone mess this `zid` business up in years, and so it's possible that we could remove this block.

karlstolley replied:
Of course, maybe this bit of documentation is _just that effective_ :) happy to remove if that seems the better route.

timabbott commented on docs/subsystems/html-css.md:
```diff
@@ -38,6 +44,10 @@ understand the current styling or modify it. We would very much like
 styling, clean up now-unused CSS, etc., to keep things maintainable.
 
+Opt to write CSS in CSS files. Avoid using the `style=` attribute in
+HTML unless the styling is actually dynamic. Instead, define logical
```
"actually dynamic" might benefit from an example, like the stream colors that depend on user data.

timabbott commented:
This generally looks like a nice set of improvements. I posted some comments; I'm generally fine with many of those being addressed in follow-up commits rather than rebasing them in (since rebasing in is pretty unfun with code moves involved).

I think probably following this PR, you/me/Anders should do a bit more of a systematic look at content that we can cut, since I only commented on items of that form that I happened to look at while reading this.

karlstolley commented:
Okay, Tim. I believe I've addressed all of the comments you've left here.

Smaller changes I implemented as rebases, but the bigger ones are now additional commits--if only to make it easier to tinker with those instead of having to rebase something a dozen commits deep.

timabbott commented on docs/contributing/code-style.md:
```diff
@@ -365,20 +365,6 @@ Additional notes:
-`rows.id(x)`. This returns a number. Even in cases where you do want a
-string, use the `id` function, as it will simplify future code changes.
-In most contexts in JavaScript where a string is needed, you can pass a
-number without any explicit conversion.
```
Just a small comment that generally I prefer to not use GitHub username mentions like `@timabbott` in commit messages because GitHub sends you annoying emails, sometimes multiple times for an individual commit as its gets added to forks, when doing so. They need a "silent mention" version of the feature :)

karlstolley replied:
Oh, that's completely annoying. Sorry about that!

gnprice replied:
We have an item for this in the zulip-mobile style guide: https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#mentioning-people

The point is generally applicable, so it'd probably be useful to put that someplace central.

timabbott commented:
This is a great improvement, merged, thanks @karlstolley!
